staging/src/k8s.io/csi-translation-lib/translate_test.go

```diff
@@ -77,6 +89,205 @@ func TestTranslationStability(t *testing.T) {
+		},
+		{
+			name:                 "AWS EBS with zone labels and topology",
+			pv:                   makeAWSEBSPV(defaultZoneLabels, makeTopology(v1.LabelZoneFailureDomain, "us-east-2a")),
+			expectedNodeAffinity: makeNodeAffinity(false /*multNodeSelectorTerms*/, plugins.AWSEBSTopologyKey, "us-east-2a"),
+		},
+		// Cinder test cases: test mosty topology key, i.e., don't repeat testing done with GCE
+		{
+			name:                 "OpenStack Cinder with zone labels",
+			pv:                   makeCinderPV(defaultZoneLabels, nil /*topology*/),
+			expectedNodeAffinity: makeNodeAffinity(false /*multNodeSelectorTerms*/, plugins.CinderTopologyKey, "us-east-1a"),
+		},
+		{
+			name:                 "OpenStack Cinder with zone labels and topology",
+			pv:                   makeCinderPV(defaultZoneLabels, makeTopology(v1.LabelZoneFailureDomain, "us-east-2a")),
+			expectedNodeAffinity: makeNodeAffinity(false /*multNodeSelectorTerms*/, plugins.CinderTopologyKey, "us-east-2a"),
+		},
+	}
+
+	for _, test := range testCases {
+		ctl := New()
+		t.Logf("Testing %v", test.name)
+
+		// Translate to CSI PV and check translated node affinity
+		newCSIPV, err := ctl.TranslateInTreePVToCSI(test.pv)
+		if err != nil {
+			t.Errorf("Error when translating to CSI: %v", err)
+		}
+
+		nodeAffinity := newCSIPV.Spec.NodeAffinity
+		if !reflect.DeepEqual(nodeAffinity, test.expectedNodeAffinity) {
+			t.Errorf("Expected node affinity %v, got %v", *test.expectedNodeAffinity, *nodeAffinity)
+		}
+
+		// Translate back to in-tree and make sure node affinity is still set
+		newInTreePV, err := ctl.TranslateCSIPVToInTree(newCSIPV)
+		if err != nil {
+			t.Errorf("Error when translating to in-tree: %v", err)
+		}
+
+		nodeAffinity = newInTreePV.Spec.NodeAffinity
+		if !reflect.DeepEqual(nodeAffinity, test.expectedNodeAffinity) {
+			t.Errorf("Expected node affinity %v, got %v", *test.expectedNodeAffinity, *nodeAffinity)
+		}
+	}
+}
+
+func makePV(labels map[string]string, topology *v1.NodeSelectorRequirement) *v1.PersistentVolume {
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels: labels,
+		},
+		Spec: v1.PersistentVolumeSpec{},
+	}
+
+	if topology != nil {
+		pv.Spec.NodeAffinity = &v1.VolumeNodeAffinity{
+			Required: &v1.NodeSelector{
+				NodeSelectorTerms: []v1.NodeSelectorTerm{
+					{MatchExpressions: []v1.NodeSelectorRequirement{*topology}},
+				},
+			},
+		}
+	}
+
+	return pv
+}
+
+func makeGCEPDPV(labels map[string]string, topology *v1.NodeSelectorRequirement) *v1.PersistentVolume {
+	pv := makePV(labels, topology)
+	pv.Spec.PersistentVolumeSource = v1.PersistentVolumeSource{
+		GCEPersistentDisk: &v1.GCEPersistentDiskVolumeSource{
+			PDName:    "test-disk",
+			FSType:    "ext4",
+			Partition: 0,
+			ReadOnly:  false,
+		},
+	}
+	return pv
+}
+
+func makeGCEPDPVMultTerms(labels map[string]string, topology *v1.NodeSelectorRequirement) *v1.PersistentVolume {
```

Comment:
I think this isn't quite right. You'll want the input to look something like:

Term[0] = zone1
Term[1] = zone2

Right now, it's:

Term[0] = zone1 or zone2
Term[1] = zone1 or zone2

Reply:
Thanks, fixed